CrossBucket push, frontend

fiskus opened the pull request:
* `CopyButton` is an action button and bucket successor's select
* CopyPackageDialog is copy-pasted from Create and Update dialog
* I've split DialogForm into several components we can reuse for Update, Create and Copy dialogs. I suspect, there will be PushDirectory dialog
* I've added some snapshot tests for small components

Hierarchy for new components:
```
PackageTree
 +- CopyButton
       "click" -> BucketsListFetcher 
                   +- Menu

 +- PackageCopyDialog
   +- PackageDialog.Error
   +- PackageDialog.Success
   +- PackageDialog.Loading
   +- DialogForm
       +- FilesInput -> Dropzone
                         +- Dropzone.Header
                         +- Dropzone.FileEntry[]
                         +- Dropzone.DropMessage
                        Dropzone.Overlay
```

nl0 reviewed (changes requested):
We're close, but there are some issues.

**The most important: there should be a switch for `copy_data` parameter.**

Less important:
1. refactoring caused package creation dialog to flicker after the manifest data gets loaded
2. manifest size should not be limited for package copy dialog (ok to backlog)
3. manifest / workflows / successors data should be cached (ok to backlog)

General note: @fiskus I think you're too eager abstracting things away (trying to generalize things that aren't too general in their essence) where composing or even inlining would be better from both conceptual and maintenance standpoints. If things look similar, it doesn't always mean they are "instances" of the same pattern we should abstract away, oftentimes this pattern is not general enough. I hope this makes some sense.

fiskus commented on catalog/app/components/Dropzone/Dropzone.js:
```diff
@@ -0,0 +1,100 @@
+
+const useStyles = M.makeStyles((t) => ({
+  dropzone: {
+    marginTop: t.spacing(2),
```
Agree. But this is not `root` class

nl0 replied:
oh i see

fiskus commented on catalog/app/components/Dropzone/Dropzone.js:
```diff
@@ -0,0 +1,100 @@
+  files,
+  onDrop,
+  overlayComponent,
+  statsComponent,
```
What do you propose for a name then? `statsInstance`?

nl0 replied:
widely used convention is `Element` suffix, e.g. `statsElement`. in case of this particular components, maybe we can omit this altogether and use just `stats` and `overlay`

fiskus commented on catalog/app/components/Dropzone/Header.js:
```diff
@@ -0,0 +1,43 @@
+        })}
+      >
+        Files
+        {children}
```
In this implementation we don't need space. But I think implementation would be better:

```
Files
{children && (
  :<span className={setMarginHere}>
    {children}
  </span>
)}
```
or
```
Files
{children && <>: {children}<>}
```

nl0 replied:
ok cool.
tho `{children && ...}` should be `{!!children && ...}`

fiskus commented on catalog/app/components/Dropzone/__snapshots__/FilesStats.spec.js.snap:
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`FilesStats should render 1`] = `null`;
```
There is already, but it is skipped because `react-intl` breaks `jest`
https://github.com/quiltdata/quilt/blob/c571c1bb004fc5b941c15424d30dbf6d934dbed6/catalog/app/components/Dropzone/FilesStats.spec.js#L12

fiskus replied:
This line causes exception https://github.com/quiltdata/quilt/blob/c571c1bb004fc5b941c15424d30dbf6d934dbed6/catalog/app/utils/string.js#L34

nl0 replied:
probably it's because it requires the Intl context. not a blocker anyways

fiskus replied:
> it requires the Intl context

this

fiskus commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,142 @@
+const BucketsListFetcher = React.forwardRef(({ bucket, children }, ref) => {
+  const s3 = AWS.S3.use()
+  const data = useData(requests.workflowsList, { s3, bucket })
+  const res = data.case({
```
I'm not familiar with how to use this API properly yet. Do you mean it can be safe to do:

```
const data = useData(…)
return children({
  ref,
  ...data,
})
```

fiskus replied:
Code from my comment doesn't work with exception `Invariant Violation: app/utils/tagged/case: must be called with an instance of type, otherwise cases must include the __ placeholder`.
So please elaborate on how can I change it

nl0 replied:
useData returns an object like `{ result, fetch, case }` ([the code](https://github.com/quiltdata/quilt/blob/master/catalog/app/utils/Data.js#L56)), you need `.result` (it's an instance of AsyncResult), e.g. `return children({ ref, ...data.result })`

fiskus commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,142 @@
+    Err: AsyncResult.Err,
+    _: R.identity,
+  })
+  return children({
```
> it's not clear why we need ref here altogether
Because `BucketsListFetcher` is a child of `M.Menu`. And `M.Menu` demand for `ref`

fiskus replied:
> second, i dont think we should patch AsyncResult objects. if you need to pass a second argument to the AsyncResult.case, you can just call children with that second argument, i.e.:

I don't understand. Will be back to this later

fiskus commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,387 @@
+
+const filesInitialValue = { existing: [] }
+
+function FilesInput({ input: { value: inputValue }, meta }) {
```
To be consistent with UpdateDialog and CreateDialog. We can reuse this component there. And then we can reuse it for further package dialogs, like for packaging directory

nl0 replied:
well, we have the Dropzone component that incapsulates most of the reusable logic and layout, but using input where we dont need it doesnt seem like a good idea, it makes things more complex by forcing us to shape the data in some unnatural way and passing props to disable some unused behavior, where we could just use the underlying Dropzone component directly, w/o tying it to the final-form's input api

fiskus commented on catalog/app/components/Dropzone/index.js:
```diff
@@ -0,0 +1,3 @@
+export { default } from './Dropzone'
```
I want to spend some time hooking this up into Create and Update dialogs, and I suppose I can leave only default export and remove Overlay and Stats.

fiskus commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,142 @@
+  )
+}
+
+const BucketsListFetcher = React.forwardRef(({ bucket, children }, ref) => {
```
I agree

I checked: it makes effect on React dev tools. `console.trace()` and `throw` have the same trace in both cases

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -52,6 +56,45 @@ async function requestPackageCopy(
+  },
+}))
+
+function CopyDataSwitcher({ input }) {
```
i think we should destructure `input` and use `onChange` and `value` directly

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -52,6 +56,45 @@ async function requestPackageCopy(
+          onChange={handleChange}
+        />
+      }
+      label="Full copy"
```
probably "Copy data" would be better (@akarve wdyt?). also, it would be nice to have some hint describing the purpose of this switch in more detail (like a question mark with a popup hint maybe)

fiskus replied:
I changed the text to `Files will be copied from "${sourceBucket}" to "${targetBucket}"` when copy-data is active, and `Files will persist in "${sourceBucket}", but links will be copied` when copy-data is inactive

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -189,17 +241,27 @@ function DialogForm({
+                  className={classes.filesSwitcher}
+                  component={CopyDataSwitcher}
+                  name="copyData"
+                  initialValue
```
i believe it should be disabled by default to prevent copying over lots of data unnecessarily

fiskus commented on catalog/app/components/Dropzone/__snapshots__/DropMessage.spec.js.snap:
```diff
@@ -0,0 +1,33 @@
+
+exports[`DropMessage should be empty when disabled 1`] = `
+<div
+  className="makeStyles-root-1 makeStyles-error-3 makeStyles-warning-2"
```
I'm not sure this is a 100% good idea. Maybe image snapshots are better.
Need to RnD on jest-styled-components: does it work with our styles at all, are our styles are deterministic.
Also without styles, it tests inner logic for components. Testing styles is UI test, screenshots could be better for that job

nl0 replied:
> Maybe image snapshots are better.

well, maybe, but i think it's a very different story

> does it work with our styles at all

i think it doesn't, bc it's for styled-components, not for JSS, that's why i said "something like [it]"

> Also without styles, it tests inner logic for components

it's not "without styles" -- it actually renders the css rulesets directly in the snapshot (see the examples in the readme of jest-styled-components)

fiskus replied:
> it's not "without styles"

Right now tests don't have styles, only class names

nl0 replied:
> > it's not "without styles"
> 
> Right now tests don't have styles, only class names

yeah, but i'm talking about jest-styled-components

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -189,17 +241,27 @@ function DialogForm({
+                <RF.Field
+                  component={FilesInput}
+                  name="files"
+                  validate={validators.nonEmpty}
```
all this logic here (validation, error handling, equality checks, handling the actual field state) is unused, bc files are not editable

nl0 commented on catalog/app/components/Dropzone/__snapshots__/DropMessage.spec.js.snap:
```diff
@@ -0,0 +1,33 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`DropMessage should be empty when disabled 1`] = `
```
you should run jest again to update the snapshots

fiskus commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,387 @@
+          />
+        ),
+        Success: (props) => (
+          <PD.DialogSuccess
```
It will be different when I add it to different forms

nl0 replied:
success message for package promotion should reflect what happens, but now it shows the same message we see after a regular package push

fiskus replied:
So, we need to change title on `DialogSuccess`. What is you proposal?

nl0 replied:
maybe something like "Package pushed to [s3://destination-bucket]() as [package-name@revision]()"

fiskus commented on catalog/app/containers/Bucket/PackageDialog/DialogError.js:
```diff
@@ -0,0 +1,81 @@
+      <M.DialogTitle>{title}</M.DialogTitle>
+
+      <M.DialogContent className={classes.content}>
+        <FormSkeleton animate={false} />
```
Sure, I will customize it when I use it on in different environments

nl0 replied:
but FormSkeleton component is already used in different forms, isnt it?

nl0 replied:
not a blocker, ok to backlog, but we shouldnt forget about that

fiskus commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,142 @@
+            ),
+          _: () => <MenuPlaceholder />,
+          Err: (error) => (
+            <M.Box px={2} py={1}>
```
It barfs on components not implementing ref forwarding

fiskus commented on catalog/app/containers/Bucket/PackageCreateDialog.js:
```diff
@@ -473,195 +476,190 @@ function PackageCreateDialog({ bucket, open, workflowsConfig, onClose, refresh }
+  return DialogState.case(
+    {
+      Form: (workflowsConfig) => (
+        <PackageCreateDialog
```
I reverted this change

fiskus commented:
@nl0 

> The most important: there should be a switch for copy_data parameter.

done

> refactoring caused package creation dialog to flicker after the manifest data gets loaded

I reverted this change back

> manifest size should not be limited for package copy dialog (ok to backlog)

will backlog

> manifest / workflows / successors data should be cached (ok to backlog)

I fixed the bug with the incorrect state (incorrect revision's hash) on the successor's page after a successful push. Will backlog caching

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,417 @@
+        label={
+          value
+            ? `Files will be copied from "${sourceBucket}" to "${targetBucket}"`
+            : `Files will persist in "${sourceBucket}", but links will be copied`
```
this seems a little counter-intuitive, i think we should keep the short label (e.g. "Copy data") and show this longer text as a hint or help text or smth

nl0 commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,145 @@
+  // FIXME: add documentation link
+  return (
+    <M.Menu anchorEl={anchorEl} onClose={onClose} open={open}>
+      <BucketsListFetcher bucket={bucket}>
```
to avoid complexities with ref forwarding etc i think we should use a hook for fetching the data here -- move the useData call from BucketsListFetcher to this component -- this will also achieve successors caching

fiskus replied:
Still need to use `<div>` as a wrapper for `M.MenuItem`s to avoid `ref` issue

nl0 replied:
i tried it locally and it works. what's exactly the issue you're having?

fiskus replied:
Console's error for Menu `ref`

I fixed it by forwarding `ref` for `MenuItem`. I removed `div`, by the way

nl0 replied:
it's strange, bc i've tried it locally (edited the code) and there were no errors. well, nvm

nl0 commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,133 @@
+    <M.Menu anchorEl={anchorEl} onClose={onClose} open={open}>
+      {data.case({
+        Ok: ({ successors }) => (
+          <div>
```
this div wrapper is unnecessary (and seems like it also causes the menu positioning bug, see the attachment; removing it fixes the bug), `MenuItem`s should be the direct children of the `Menu`

<img width="533" alt="Screenshot 2020-12-01 at 20 46 09" src="https://user-images.githubusercontent.com/122294/100763069-88901180-3416-11eb-973c-a5b1c01ea7b4.png">

fiskus replied:
Yeah, thanks. When I saw it I didn't realize it's a bug, because behavior is similar to `M.Select` :)

nl0 commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  return xs ? (
+    <M.IconButton aria-haspopup="true" edge="end" title={children} {...props}>
```
`aria-haspopup="true"` is already present in `props`

nl0 commented on catalog/app/containers/Bucket/CopyButton.js:
```diff
@@ -0,0 +1,133 @@
+    size: 'small',
+  }
+
+  return xs ? (
```
probably should shift this to `sm`, bc there's not enough space for breadcrumbs even on ipad in portrait mode:
<img width="624" alt="Screenshot 2020-12-01 at 20 52 32" src="https://user-images.githubusercontent.com/122294/100763699-57fca780-3417-11eb-9e6f-0471804d3d60.png">

fiskus replied:
As well as "Download package" button?

nl0 replied:
yep

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -58,6 +58,11 @@ const useCopyDataSwitcherStyles = M.makeStyles((t) => ({
+  icon: {
+    color: t.palette.text.hint,
+    marginLeft: t.spacing(1),
+    verticalAlign: '-4px',
```
oh wow i didnt realize this is allowed

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,435 @@
+  return (
+    <div className={classes.root}>
+      <M.FormControlLabel
+        control={<M.Switch color="primary" checked={value} onChange={handleChange} />}
```
let's use a checkbox here: https://uxmovement.com/buttons/when-to-use-a-switch-or-checkbox/

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,435 @@
+      <M.Tooltip
+        title={
+          value
+            ? `Files will be copied from "${sourceBucket}" to "${targetBucket}"`
```
this may be incorrect for some cases, because files may be in different bucket(s), not necessarily in sourceBucket. @akarve pls wordsmith
```suggestion
            ? `Files will be copied to s3://${targetBucket}`
```

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,435 @@
+        title={
+          value
+            ? `Files will be copied from "${sourceBucket}" to "${targetBucket}"`
+            : `Files will persist in "${sourceBucket}", but links will be copied`
```
similar to the above. @akarve pls wordsmith
```suggestion
            : `Files from the package will be reused, only links will be copied`
```

nl0 commented on catalog/app/utils/workflows.spec.js:
```diff
@@ -189,15 +189,72 @@ describe('utils/workflows', () => {
 
-      it('should return copyData param', () => {
-        expect(config.copyData).toBe(true)
+      it('should return false by default', () => {
```
```suggestion
      it('should return true by default', () => {
```

nl0 replied:
let's also add a test for `'bucket-copy-default'`

nl0 replied:
ok, i see that it's handled by another test

nl0 commented on catalog/app/utils/workflows.js:
```diff
@@ -38,6 +50,7 @@ function parseWorkflow(workflowSlug, workflow, data) {
 }
 
 const parseSuccessor = (url, successor) => ({
+  copyData: successor.copy_data || false,
```
```suggestion
  copyData: successor.copy_data != null ? successor.copy_data : COPY_DATA_DEFAULT,
```

fiskus replied:
I'm comparing with undefined

nl0 replied:
ok, `x == null` is basically a shorthand for `x === null || x === undefined`

nl0 commented on catalog/app/utils/workflows.spec.js:
```diff
@@ -189,15 +189,72 @@ describe('utils/workflows', () => {
+            name: 'Copy default',
+            slug: 'bucket-copy-default',
+            url: 's3://bucket-copy-default',
+            copyData: false,
```
```suggestion
            copyData: true,
```

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,399 @@
+        workflow: PD.getWorkflowApiParam(workflow.slug),
+      },
+    })
+  } catch (e) {
```
i think you should not catch errors here, bc it breaks error handling in `onSubmit`

nl0 replied:
(and it doesnt work anyways bc promise returned by `req` isnt awaited)

nl0 commented on catalog/app/containers/Bucket/PackageCopyDialog.js:
```diff
@@ -0,0 +1,399 @@
+import * as PD from './PackageDialog'
+import * as requests from './requests'
+
+async function requestPackageCopy(
```
async is unnecessary here, bc you return Promise anyways and dont use `await` in the body

nl0 reviewed (changes requested):
there's a possible bug related to error handling and async / await in `onSubmit` / `requestPackageCopy`, the rest looks good
